StatsdClient/AsyncLock.cs

```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace StatsdClient
+{
+    internal sealed class AsyncLock
```

Comment:
I almost wonder if we should be using the async lock from: https://github.com/StephenCleary/AsyncEx The only downside is now that you have to have a dependency on that.. As long as this lock works I think we should be fine. Just wish it was tested.